### PR TITLE
Add email opt-in modal to homepage flow

### LINKED ISSUE #1043: Implement email address collection dialog
Blocked by https://github.com/mozilla/testpilot/issues/1074
- [ ] Create email signup modal for optional sign up after installation.
- [ ] Move basket integration to email signup modal.

### PR DESCRIPTION
**Pull request by jaredhirsch**
Fixes #1043.
- Set a cookie on addon install click, then on the experiment list page, if the cookie's set, unset it and show the modal.
- Basket is all wired up (I'm getting like 20 test pilot newsletters ;-)
- pull in HTML5 email validation rules using `micro-email-validator`

<strike>Looks like the image is missing from the commit, I'll fix that later today.</strike> _(Edit: image added)_

### CODE REVIEW

**Comment by meandavejustice** [comment]
Will review this in the morning, gotta stop looking at my computer

File: testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      eventCategory: 'HomePage Interactions',
+      eventAction: 'button click',
+      eventLabel: 'On to the experiments'
+    });
+
+    this._close();
+  },
+
+  validate(email) {
+    return emailValidator(email);
```
Comment (ckprice) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:129]: `bob@bob` passed this validator. Looks like basket threw an error though.

email validator

**Comment by ckprice** [comment]
Eh, didn't see Dave assign himself to this, so you can take this fwiw. I gave the functionality a looking at and confirmed the following
- the modal appears after install.
- all events are being sent.
- valid email addresses are submitting to basket, and receipt of welcome email.
- didn't get modal on subsequent visits.

Left a comment re: validator.

File: testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js (hunk 1 above)
Comment (jaredhirsch) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:129]: > `bob@bob` passed this validator.

@ckprice Yeah, email validation is hard. The validator library just copies Firefox's implementation of HTML5 email constraint validation, same approach as FxA. This means that some silly-looking emails will be counted as valid, like `root@localhost`. These are technically valid, according to the email RFC and the HTML5 validation spec.

I'm happy to explore other options, here or in a followup bug; let me know if you have a particular validation library in mind.

**Comment by jaredhirsch** [comment]
@ckprice Thanks for taking a look :beers:

File: testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js (hunk 1 above)
Comment (ckprice) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:129]: I think we should do our best to allow what basket is allowing. This [production snippet used RegEx](https://github.com/mozmar/snippets/blob/761df2f2a9b6f977c710d44078b27b3a7edd1da5/campaigns/mobile/mobile-snippet-widget.html#L379) to validate email addresses to basket.

Comment (jaredhirsch) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:129]: cool, sounds good

File: testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    this.set('isValidEmail', true);
+
+    app.sendToGA('event', {
+      eventCategory: 'HomePage Interactions',
+      eventAction: 'button click',
+      eventLabel: 'Sign me up'
+    });
+
+    app.subscribeToBasket(email, () => {
+      // TODO: review question: basket failures are swallowed by the subscribeToBasket
```
Comment (ckprice) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:87]: I would vote no. I don't see any examples of this in the existing apps using basket.

File: testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  },
+
+  skip(e) {
+    e.preventDefault();
+    e.stopPropagation();
+
+    app.sendToGA('event', {
+      eventCategory: 'HomePage Interactions',
+      eventAction: 'button click',
+      // review TODO: The label says 'skip'. Should I just use that? Want to make
```
Comment (ckprice) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:108]: +1 to `Skip email`

File: testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  events: {
+    'click button[data-hook=submit-email]': 'submit',
+    'click [data-hook=cancel-modal]': 'skip',
+    'click [data-hook=continue]': 'continue'
+  },
+
+  submit(e) {
+    e.preventDefault();
+
+    // TODO: should we log the number of email validity failures? worth tracking?
```
Comment (ckprice) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:70]: I think this number is interesting in the general "how people use forms" aspect, but probably not super valuable to us.

**Comment by ckprice** [comment]
This looks good to me with a tiny comment. I'll give @meandavejustice the chance to have a look if he wants before I merge though.

**Comment by lmorchard** [comment]
Taking a peek

**Comment by lmorchard** [comment]
Looks good to me

File: testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js (hunk 1 above)
Comment (jaredhirsch) [inline testpilot/frontend/static-src/app/views/email-opt-in-dialog-view.js:129]: This got merged before I had a chance to swap out the validator, I'll open a followup PR for that

**Pull request merged**